File: internal/app/annotate.go
package app

import (
	"github.com/leandronsp/githerb/internal/review"
)

// Annotate leaves a comment on a range of lines of a proposal's head revision.
// This is the human half of the loop.
type Annotate struct {
	Proposals review.Proposals
	Author    string
	Now       Clock
}

// Run appends the comment and returns it.
func (a Annotate) Run(id, file, side string, start, end int, body string) (review.Comment, error) {
	proposal, err := a.Proposals.Load(review.ProposalID(id))
	if err != nil {
		return review.Comment{}, err
	}

	parsed, err := review.ParseSide(side)
	if err != nil {
		return review.Comment{}, err
	}

	span, err := review.NewSpan(parsed, start, end)
	if err != nil {
		return review.Comment{}, err
	}

	head := proposal.Head().SHA()

	comment, err := review.NewComment(head, review.File(file), span, body, a.Author, a.Now())
	if err != nil {
		return review.Comment{}, err
	}

	// Ask the aggregate first: it refuses a comment that does not belong here.
	if _, err := proposal.WithRecord(review.CommentRecord(comment)); err != nil {
		return review.Comment{}, err
	}

	if err := a.Proposals.Annotate(head, review.CommentRecord(comment)); err != nil {
		return review.Comment{}, err
	}

	return comment, nil
}

// Resolve marks a comment as dealt with. This is the agent half of the loop.
type Resolve struct {
	Proposals review.Proposals
	Author    string
	Now       Clock
}

// Run appends the resolution.
func (r Resolve) Run(id, comment string) error {
	proposal, err := r.Proposals.Load(review.ProposalID(id))
	if err != nil {
		return err
	}

	resolution, err := review.NewResolution(review.ID(comment), r.Author, r.Now())
	if err != nil {
		return err
	}

	if _, err := proposal.WithRecord(review.ResolutionRecord(resolution)); err != nil {
		return err
	}

	return r.Proposals.Annotate(proposal.Head().SHA(), review.ResolutionRecord(resolution))
}

File: crates/app/src/annotate.rs
//! Leaving a note on a range of lines. This is the human half of the loop.

use review::{Anchor, Author, Comment, ProposalId, Record, Timestamp};

use crate::error::Result;
use crate::store::Store;

/// Write a note against the head revision of a proposal.
///
/// The anchor arrives already parsed, because a file and a span are what the
/// core understands and a surface that has not turned its strings into one
/// has not finished reading its input. The aggregate is asked next, because
/// it is what knows whether the note belongs here at all: a line that would
/// not fold is a line that never reaches the log.
///
/// # Errors
///
/// A proposal nobody opened, one that is no longer open, or a body that says
/// nothing.
pub fn annotate(
    store: &Store,
    author: &Author,
    now: Timestamp,
    id: &ProposalId,
    anchor: Anchor,
    body: &str,
) -> Result<Comment> {
    let mut proposal = store.load(id)?;
    let head = proposal.head().sha().clone();

    let comment = Comment::new(head.clone(), anchor, body, author.clone(), now)?;
    let record = Record::Comment(comment.clone());

    proposal.apply(record.clone())?;
    store.annotate(&head, &record)?;

    Ok(comment)
}
